services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/tenant/ManagementStore.java

```diff
@@ -443,4 +453,41 @@ protected Future<UpdateResult> updateJsonField(
                 checkSpan -> readTenantEntryById(this.client, tenantId, checkSpan.context()));
     }
 
+    /**
+     * Gets a list of tenants.
+     * @param pageSize the page size
+     * @param pageOffset the page offset
+     * @param spanContext The span to contribute to.
+     *
+     * @return A future containing tenants
+     */
+    public Future<SearchResult<TenantWithId>> find(final int pageSize, final int pageOffset, final SpanContext spanContext) {
+
+        final var expanded = this.getTenantsStatement.expand(map -> {
+            map.put("page_size", pageSize);
+            map.put("page_offset", pageOffset); });
+
+        final Span span = tracer.buildSpan("find Tenants")
+            .addReference(References.CHILD_OF, spanContext)
+            .start();
+
+        return expanded
+            .trace(this.tracer, span.context())
+            .query(this.client)
+            .map(r -> {
+                final var entries = r.getRows(true);
+                span.log(Map.of(
+                    "event", "read result",
+                    "rows", entries.size()));
+                final List<TenantWithId> list = new ArrayList<>();
+                for (var entry : entries) {
+                    final var id = entry.getString("tenant_id");
+                    final var tenant = Json.decodeValue(entry.getString("data"), Tenant.class);
+                    final var version = Optional.ofNullable(entry.getString("version"));
+                    list.add(TenantWithId.from(id, tenant));
+                }
+                final SearchResult<TenantWithId> searchResult = new SearchResult<>(list.size(), list);
```

Comment:
IMHO using the requested page size as the overall number of matching tenants will not yield a proper search result because the client will always get the impression that there are no tenants in addition to the page that he has just retrieved. FMPOV there needs to be COUNT query involved somewhere which determines the overall number of tenants, or am I mistaken?

I guess the same is true for the device search ...

Reply:
Thanks for bringing this up, I was actually wondering how to interpret the documentation...
In the current implementation the returned number of items is only equal to the page size if there are at least pageSize elements found in total by the query. Otherwise it is less. In the documentation it says "The total number of objects in the result set, regardless of the *pageSize* set in query."
My understanding of the documentation was to report the actual number of returned items instead of the overall number of items that would be found without page limitation. Most APIs I know of are not returning a total number of items, but instead do return an indication that there are more items, e.g. by returning a nextPage token, see for example the corresponding IoT Core API: https://cloud.google.com/iot/docs/reference/cloudiot/rest/v1/projects.locations.registries/list
I don't see a big benefit in returning the total number of items, because a client could just continue querying as long as the number of returned items is equal to pageSize. Only for the very last call it might occasionally happen, that it gets an empty list because all items already have been returned with the previous call.
Adding an additional COUNT query would also make all API calls considerably slower.
If you think it's still necessary, e.g. due to compatibility reasons with the MongoDB implementation, I can of course add it.

Reply:
> My understanding of the documentation was to report the actual number of returned items instead of the overall number of items that would be found without page limitation.

No, it is the latter. 

> I don't see a big benefit in returning the total number of items, because a client could just continue querying as long as the number of returned items is equal to pageSize. Only for the very last call it might occasionally happen, that it gets an empty list because all items already have been returned with the previous call.
Adding an additional COUNT query would also make all API calls considerably slower.
If you think it's still necessary, e.g. due to compatibility reasons with the MongoDB implementation, I can of course add it.

I understand your point and I agree that there is room for improvement (as always ;-))
However, the current API is designed differently and therefore I would kindly ask you to change your code accordingly. It might be true that doing an additional COUNT query increases the work that needs to be done. On the other hand, this is the management API and we do not expect this API to be invoked in the thousands of requests per second, so I do not expect any issues here.

With the next major release of Hono we can change this API (in a breaking way) if we think that it is beneficial.

Reply:
done